### PR TITLE
Docs: Add ux-guidelines document

### LINKED ISSUE #27344: Compiler UI/UX Design Guidelines Document
There's a lot of compiler UI and UX rules that exist in the developer mindshare, but are not documented anywhere. They aren't really Rust code style guidelines, since they only apply to the compiler (and possibly compiler plugins), so we really need to have another document for them.

A couple examples for this are:

* Error messages **MUST NOT** use `illegal` when `invalid` or similar works. See issue #27288.
* `Error` and `Warning` messages **SHOULD NOT** suggest how to fix them. If there is a common fix, they should be done with a `Note`.

### PR DESCRIPTION
**Pull request by Havvy**
r? @steveklabnik 

Closes #27344.

### CODE REVIEW

File: src/doc/ux-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+These guidelines are for the rustc compiler.
+
+## Error Messages
+
+* The word "illegal" is illegal. Prefer "invalid" or a more specific word
+instead.
+* Errors and Warnings should not suggest how to fix the problem. A Help
+message should be emitted instead.
+* Errors, Warnings, Notes, and Help messages start with a lowercase letter and
+not end with punctuation.
```
Comment (GuillaumeGomez) [inline src/doc/ux-guidelines.md]: Missing "do" -> "do not end..."

**Comment by Gankro** [comment]
This will show up at http://doc.rust-lang.org/ux-guidelines.html, yes?

**Comment by Havvy** [comment]
@nagisa said on IRC that it'll show up there, but nothing will link to it.

@Manishearth was thinking a % sign was needed at the top to be rendered.

I honestly have no clue. I asked in the initial bug where this should go, deciding on this location unless somebody tells me differently.

**Comment by Gankro** [comment]
Yes rustdoc expects the first heading to be a `%` instead of a `#` -- it will become the title of the page.

**Comment by steveklabnik** [comment]
So, I am generally in favor, but I think we want to put these suggestions through the RFC process first. If I were to just merge this, it would be unilaterally declaring a style, you know?

What do you think, @aturon

**Comment by Havvy** [comment]
I've started a pre-RFC @ https://internals.rust-lang.org/t/pre-rfc-rustc-ux-guidelines/2419 - It's letting me get more points added to the document in any case.

Note though, that ux guidelines have been changed without RFC, even recently with the `illegal -> invalid` changes that happened this week. And this document isn't actually making any tough decisions - it's documenting decisions that were already (implicitly) made.

Since I'm starting to get suggestions from other people though, I need to make changes to this, so it's not good to merge. As such, closing for now.

**Pull request closed without merging**